File: LeetCode/balanced_binary_tree.py
#!/usr/bin/env python
#coding: utf-8

## Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def setHeight(self, root):
        if not root: return 0
        leftH = self.setHeight(root.left)
        rightH = self.setHeight(root.right)
        root.val = max(leftH, rightH) + 1
        return root.val
    
    def _isBalanced(self, root):
        if not root: return True
        if not root.left and not root.right: return True
        if not root.left and root.right: return root.right.val == 1
        if root.left and not root.right: return root.left.val == 1

        return (self._isBalanced(root.left) and self._isBalanced(root.right)
                and (abs(root.left.val - root.right.val) <= 1))

    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        self.setHeight(root)
        return self._isBalanced(root)

if __name__ == '__main__':
    s = Solution()
    root = TreeNode(1)
    root.right = TreeNode(2)
    assert True == s.isBalanced(root)

File: LeetCode/Java/Solution/src/ios/ac/cn/BalancedBinaryTree.java
package ios.ac.cn;

public class BalancedBinaryTree {
	/**
	 * If root is a balanced tree, return it's height, 
	 * or return -1
	 * 
	 * @param root
	 * @return
	 */
	private int getHeight(TreeNode root) {
		if (root == null) {
			return 0;
		}

		int leftHeight = getHeight(root.left);
		if (leftHeight == -1) {
			return -1;
		}

		int rightHeight = getHeight(root.right);
		if (rightHeight == -1) {
			return -1;
		}

		if (Math.abs(leftHeight - rightHeight) <= 1) {
			return Math.max(leftHeight, rightHeight) + 1;
		} else {
			return -1;
		}
	}

	public boolean isBalanced(TreeNode root) {
		int height = getHeight(root);
		if (height >= 0) {
			return true;
		} else {
			return false;
		}
	}
}
